mopidy/models.py

```diff
@@ -2,30 +2,156 @@
 
 import json
 
+# TODO: split into base models, serialization and fields?
+
+
+class Field(object):
+    def __init__(self, default=None, type=None, choices=None):
+        """
+        Base field for use in :class:`ImmutableObject`. These fields are
+        responsible type checking and other data sanitation in our models.
+
+        For simplicity fields use the Python descriptor protocol to store the
+        values in the instance dictionary. Also note that fields are mutable if
+        the object they are attached to allow it.
+
+        Default values will be validated with the exception of :class:`None`.
+
+        :param default: default value for field
+        :param type: if set the field value must be of this type
+        :param choices: if set the field value must be one of these
+        """
+        self._name = None  # Set by FieldOwner
+        self._choices = choices
+        self._default = default
+        self._type = type
+
+        if self._default is not None:
+            self.validate(self._default)
+
+    def validate(self, value):
+        """Validate and possibly modify the field value before assignment"""
+        if self._type and not isinstance(value, self._type):
+            raise TypeError('Expected %s to be a %s, not %r' %
+                            (self._name, self._type, value))
+        if self._choices and value not in self._choices:
+            raise TypeError('Expected %s to be a one of %s, not %r' %
+                            (self._name, self._choices, value))
+        return value
+
+    def __get__(self, instance, owner):
+        if not instance:
+            return self
+        return instance.__dict__.get(self._name, self._default)
+
+    def __set__(self, instance, value):
+        if value is not None:
+            value = self.validate(value)
+
+        if value is None or value == self._default:
+            self.__delete__(instance)
+        else:
+            instance.__dict__[self._name] = value
+
+    def __delete__(self, instance):
+        instance.__dict__.pop(self._name, None)
+
+
+class String(Field):
+    def __init__(self, default=None):
+        """
+        Specialized :class:`Field` which is wired up for bytes and unicode.
```

Comment:
I happened to learn about `intern` today, and doing a quick test on ~14k tracks from a json library I found I can shrink the memory footprint of those models from ~75MB to ~55MB by interning `uri`, `musicbrainz_id` and `date` strings. The catch being that this only works for `bytes` not `unicode`, and of course in Python 3 it's the other way around... But might be a worth revisiting as for a couple of lines added it helped quite a lot. Though on the other side, if we stop duplicating artists and albums the gains would be even larger even without this.